h/services/list_groups.py

```diff
@@ -41,36 +41,74 @@ def _authority(self, user=None, authority=None):
         Return a list of groups relevant to this session/profile (i.e. user).
 
         Return a list of groups filtered on user and authority. All open
         groups matching the authority will be included.
         """
         all_open_groups = self._open_groups(user, authority)
-        private_groups = self._private_groups(user)
+        user_groups = self._user_groups(user)
+
+        return all_open_groups + user_groups
+
+    def session_groups(self, authority, user=None):
+        """
+        Return a list of groups relevant to the user-session combination,
+        in this order:
+
+        - WORLD GROUP:
+          The special world group is returned if `authority` is the default
+          authority
+        - ALL USER GROUPS:
+          "User groups" here means any group that the user is a member of:
+          this can include both private and restricted groups.
+
+        This will return all groups that the session's user is a member of
+        regardless of group type or scope. No open groups are returned.
+
+        This should return the list of groups that is appropriate for
+        activity pages and/or other views on the h service.
+        """
 
-        return all_open_groups + private_groups
+        world_group = self._world_group(authority)
+        world_group = [world_group] if world_group else []
+        user_groups = self._user_groups(user)
+
+        return world_group + user_groups
 
     def request_groups(self, authority, user=None, document_uri=None):
         """
-        Return a list of groups relevant to this request and user combination.
+        Return a list of groups relevant to this request context.
 
         Return a list of groups filtered on user, authority, document_uri.
-        Include all types of relevant groups (open and private).
-
-        Open groups will be filtered by scope (via document_uri).
+        Groups are returned in this order:
+
+        - OPEN AND RESTRICTED GROUPS:
+          Only those open or restricted group that match scope of document_uri
+          will be returned (if document_uri is missing, no open or restricted
+          groups will be returned)
+        - WORLD GROUP:
+          The special world group is returned if `authority` is the default
+          authority
+        - PRIVATE GROUPS:
+          All private groups for the user will be returned
+
+          This should return a list of groups appropriate to the client
+          via the API.
         """
-        scoped_open_groups = self._scoped_open_groups(authority, document_uri)
+        scoped_groups = self._scoped_groups(authority, document_uri)
 
         world_group = self._world_group(authority)
         world_group = [world_group] if world_group else []
 
         private_groups = self._private_groups(user)
 
-        return scoped_open_groups + world_group + private_groups
+        return scoped_groups + world_group + private_groups
```

Comment:
The code for this and `session_groups()` above is very nice! Very short and simple. I love how they encapsulate just the logic of what lists of groups constitute the session groups and the request groups, and nothing else